explorer/interpreter/type_checker.h

```diff
@@ -393,6 +393,12 @@ class TypeChecker {
                        Nonnull<const Value*> actual,
                        const ImplScope& impl_scope) const -> ErrorOr<Success>;
 
+  // Implementation of Substitute. Does not check that bindings are nonempty,
+  // nor does it trace its progress.
```

Comment:
How about instead Substitute -> SubstituteAndVerify, SubstituteImpl -> Substitute? Calling it "SubstituteImpl" seems fine for the self-recursion, a little more unusual when it's also being called from the rewrite class.